pallene/checker.lua

```diff
@@ -451,9 +452,70 @@ function Checker:check_stat(stat)
             self:add_local(stat.decl)
             self:check_stat(stat.block)
         end)
+    elseif tag == "ast.Stat.ForIn" then
+        local rhs = stat.exps
+        self:expand_function_returns(3, rhs)
+
+        if not rhs[1] then
+            type_error(stat.loc, "missing right hand side of for-in loop")
+        end
+
+        if not rhs[2] then
+            type_error(rhs[1].loc, "missing state variable in for-in loop")
+        end
+
+        if not rhs[3] then
+            type_error(rhs[1].loc, "missing control variable in for-in loop")
+        end
+
+        local expected_ret_types = {}
+        for _ = 1, #stat.decls do
+            table.insert(expected_ret_types, types.T.Any())
+        end
+
+        local itertype = types.T.Function({ types.T.Any(), types.T.Any() }, expected_ret_types)
+        rhs[1] = self:check_exp_synthesize(rhs[1])
+        local iteratorfn = rhs[1]
+
+
+        if not self:verify_iterator(iteratorfn._type, itertype) then
+            type_error(iteratorfn.loc, "expected %s but found %s in loop iterator",
+                types.tostring(itertype), types.tostring(iteratorfn._type))
+        end
+
+        rhs[2] = self:check_exp_synthesize(rhs[2])
+        rhs[3] = self:check_exp_synthesize(rhs[3])
+
+        if rhs[2]._type._tag ~= "types.T.Any" or
+           rhs[3]._type._tag  ~= "types.T.Any" then
+            type_error(rhs[2].loc, "expected state and control values to have type any")
+        end
```

Comment:
Could we have separate type checks for the state and for the control variable? That way the error message will inform which of them was the one that had a problem. It would also be nice if the error message also said the actual type as well.

Reply:
True, I'll change it to something like `expected <type> but found <type> in for-in state value`.